### PR TITLE
Handle scrape cache better for failed scrapes

### PR DESCRIPTION
**Pull request by brian-brazil**
@brancz This should let us have a much more correct idea of churn, though the exhaustion mitigation messes that up a little.

### CODE REVIEW

File: scrape/scrape.go [hunk 1]
```diff
@@ -1184,7 +1185,7 @@ loop:
 		return total, added, err
 	}
 
-	sl.cache.iterDone()
+	sl.cache.iterDone(len(b) > 0)
```
Comment (brancz) [inline scrape/scrape.go]: this deserves a comment explaining why this condition makes it right to drop cache, also extracting it here into a variable that makes it obvious already what this condition does would be helpful as well

File: scrape/scrape.go [hunk 2]
```diff
@@ -643,28 +646,44 @@ func newScrapeCache() *scrapeCache {
...
-		if c.iter-e.lastIter > 2 {
-			delete(c.series, s)
+func (c *scrapeCache) iterDone(flushCache bool) {
+	c.metaMtx.Lock()
+	count := len(c.series) + len(c.droppedSeries) + len(c.metadata)
+	c.metaMtx.Unlock()
+
+	if flushCache {
+		c.successfulCount = count
+	} else if count > c.successfulCount*2+100 {
```
Comment (brancz) [inline scrape/scrape.go]: My understanding is that this is a heuristic, that aims for tolerating targets that drop and replace all metrics in a consecutive scrape plus the metadata (which should be somewhat consistent in either case). At the very least this needs a comment describing the heuristic, but I'd also like us to consider different heuristics. For example as I said metadata is probably irrelevant, at least in terms of churn. So maybe just x2 or x3, without metadata is simpler and should allow us to catch close to all churn situations. Please correct me if I misunderstood any of this, which may very well have happened.

Comment (brian-brazil) [reply scrape/scrape.go]: This is more that if there's a small bit of churn don't lose anything. The 100 is to somewhat handle the case where this happens before there's been a successful scrape.

Metadata churn is a possibility. We've never seen it, but may aswell protect against it while we're at it.

File: scrape/scrape.go (hunk 1 above)
Comment (brian-brazil) [reply scrape/scrape.go]: Done

File: scrape/scrape.go [hunk 3]
```diff
@@ -643,28 +646,44 @@ func newScrapeCache() *scrapeCache {
...
+		// If a target had varying labels in scrapes that ultimately failed,
+		// the caches would grow indefinitely. Force a flush when this happens.
```
Comment (brancz) [inline scrape/scrape.go]: Force flushing should probably have a metric for one to be able to track if this happens often.

File: scrape/scrape.go (hunk 2 above)
Comment (brancz) [reply scrape/scrape.go]: I’d like us to at least document the heuristic and maybe explain that flushing the cache on initial iter is ok (as successfulCount will still be 0 and count is likely to be higher than 100)

Comment (brian-brazil) [reply scrape/scrape.go]: Done, and bumped it to 1k.

File: scrape/scrape.go (hunk 3 above)
Comment (brian-brazil) [reply scrape/scrape.go]: Done

File: scrape/scrape_test.go [hunk 4]
```diff
@@ -779,6 +779,146 @@ func TestScrapeLoopRunCreatesStaleMarkersOnParseFailure(t *testing.T) {
...
+	}
+
+	// 1 successfully scraped sample, 1 stale marker after first fail, 4 report samples for
+	// each scrape successful or not.
+	if len(appender.result) != 22 {
+		t.Fatalf("Appended samples not as expected. Wanted: %d samples Got: %d", 22, len(appender.result))
+	}
+}
+
+func TestScrapeLoopCacheMemoryExhaustionProtection(t *testing.T) {
```
Comment (brancz) [inline scrape/scrape_test.go]: This test seems to fail consistently on CI

Comment (brian-brazil) [reply scrape/scrape_test.go]: I had changed the constant without updating the test, should work now.

**Comment by brancz** [comment]
lgtm

**Pull request merged**